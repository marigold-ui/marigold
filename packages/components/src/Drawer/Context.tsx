import { createContext, useContext } from 'react';
import { ThemeComponentParts } from '@marigold/system';

export const DrawerContext = createContext<{
  classNames: { [Part in ThemeComponentParts<'Drawer'>]: string };
}>({
  classNames: {
    closeButton: '',
    container: '',
    header: '',
    content: '',
    actions: '',
  },
});
export const useDrawerContext = () => useContext(DrawerContext);
