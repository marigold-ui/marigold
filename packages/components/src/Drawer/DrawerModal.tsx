import { cn, useSmallScreen } from '@marigold/system';
import { Modal, NonModal } from '../Overlay';
import type { NonModalProps } from '../Overlay/NonModal';

// Props
// ---------------
export type DrawerModalProps = NonModalProps;

// Component
// ---------------
export const DrawerModal = ({
  children,
  className,
  ...props
}: DrawerModalProps) => {
  const isSmallScreen = useSmallScreen();

  return isSmallScreen ? (
    <Modal>{children}</Modal>
  ) : (
    <NonModal
      {...props}
      className={cn('fixed top-0 right-0 bottom-0', className)}
    >
      {children}
    </NonModal>
  );
};
