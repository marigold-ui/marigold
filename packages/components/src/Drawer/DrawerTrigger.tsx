import type RAC from 'react-aria-components';
import { DialogTrigger } from 'react-aria-components';

// Props
// ---------------
export interface DrawerTriggerProps
  extends Omit<RAC.DialogTriggerProps, 'isOpen'> {
  /**
   * Whether the overlay is open by default (controlled).
   * @default false
   */
  open?: boolean;
}

// Component
// ---------------
export const DrawerTrigger = ({
  open,
  children,
  ...props
}: DrawerTriggerProps) => (
  <DialogTrigger isOpen={open} {...props}>
    {children}
  </DialogTrigger>
);
