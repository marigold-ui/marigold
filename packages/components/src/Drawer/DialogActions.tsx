import type { CSSProperties, ReactNode } from 'react';
import { cn, useClassNames } from '@marigold/system';
import { useDrawerContext } from './Context';

export interface DrawerActionsRenderProps {
  close: () => void;
}

export interface DrawerActions {
  variant?: string;
  size?: string;
  /**
   * Children of the component.
   */
  children?: ReactNode;
}

export const DrawerActions = ({ variant, size, children }: DrawerActions) => {
  const ctx = useDrawerContext();
  const classNames = useClassNames({
    component: 'Drawer',
    variant: variant ?? ctx.variant,
    size: size ?? ctx.size,
  });

  return (
    <div
      className={cn('[grid-area:actions]', classNames.actions)}
      style={{ '--i': 2 } as CSSProperties}
    >
      {children}
    </div>
  );
};
