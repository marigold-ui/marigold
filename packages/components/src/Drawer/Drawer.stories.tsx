import type { Meta, StoryObj } from '@storybook/react';
import { useState } from 'react';
import { Button } from '../Button';
import { Checkbox } from '../Checkbox';
import { Select } from '../Select';
import { Slider } from '../Slider';
import { Stack } from '../Stack';
import { TextField } from '../TextField';
import { Drawer, DrawerProps } from './Drawer';

const meta = {
  title: 'Components/Drawer',
  component: Drawer,
  argTypes: {
    placement: {
      control: {
        type: 'radio',
      },
      description: 'The placement of the drawer on the screen.',
      options: ['right', 'left', 'top', 'bottom'],
    },
    size: {
      control: {
        type: 'radio',
      },
      description: 'The size of the drawer on the screen.',
      options: ['default', 'xsmall', 'small', 'medium'],
    },
  },
  args: {},
} satisfies Meta<DrawerProps>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {
  render: args => (
    <Stack space={8} alignX="left">
      <Drawer.Trigger>
        <Button>Open Drawer</Button>
        <Drawer {...args}>
          <Drawer.Title>Drawer Title</Drawer.Title>
          <Drawer.Content>
            <p className="pb-4">
              Once upon a time in the quirky world of web design, there lived a
              lively Drawer Component named Daria. Unlike her staid companions,
              she loved to make a grand entrance by sliding in from the left,
              declaring, "Surprise! I've got filters and puns galore!"
            </p>

            <p className="pb-4">
              Every time a user clicked, Daria would pop out with flair, showing
              off her hidden treasures while her neighboring components—Header
              the Bold and Footer the Wise—watched in amused awe. They’d chuckle
              as Daria’s witty remarks filled the space between content
              sections.
            </p>

            <p className="pb-4">
              One day, Daria got a bit too excited. In a burst of enthusiasm,
              she slid out so fast that she disrupted the perfect grid layout!
              The Main Content, usually so composed, was thrown into disarray,
              muttering about lost pixels and wayward margins.
            </p>

            <p className="pb-4">
              But instead of sulking, Daria quipped, "Looks like I just broke
              the grid—guess I'm too stylish to be confined!" The users,
              entertained by her antics, cheered on as she restored order with a
              graceful slide back, proving that even a mischievous drawer could
              be the hero of responsive design.
            </p>
          </Drawer.Content>
          <Drawer.Actions>
            <Button slot="close">Close</Button>
            <Button
              slot="close"
              variant="primary"
              onPress={() => alert('Drawer will be closed')}
            >
              Save
            </Button>
          </Drawer.Actions>
        </Drawer>
      </Drawer.Trigger>
      <TextField
        label="Name"
        description="Can you interact with me when the drawer is open?"
        width={80}
      />
    </Stack>
  ),
};

export const WithForms: Story = {
  render: args => (
    <Drawer.Trigger>
      <Button>Configure Filter</Button>
      <Drawer {...args}>
        <Drawer.Title>Filter</Drawer.Title>
        <Drawer.Content>
          <Stack space={8}>
            <Slider
              label="Price"
              formatOptions={{ style: 'currency', currency: 'EUR' }}
              minValue={10}
              maxValue={140}
              defaultValue={[30, 60]}
              thumbLabels={['min', 'max']}
            />
            <Select label="Category">
              <Select.Option id="all">All</Select.Option>
              <Select.Option id="classic">Classic</Select.Option>
              <Select.Option id="rock">Rock</Select.Option>
              <Select.Option id="pop">Pop</Select.Option>
              <Select.Option id="jazz">Jazz</Select.Option>
            </Select>
            <Checkbox.Group label="Amenities">
              <Checkbox label="Fast Lane" value="fast-lane" />
              <Checkbox label="VIP Parking" value="parking" />
            </Checkbox.Group>
          </Stack>
        </Drawer.Content>
        <Drawer.Actions>
          <Button slot="close">Close</Button>
          <Button
            slot="close"
            variant="primary"
            onPress={() => alert('Apply filters and close dialog')}
          >
            Apply
          </Button>
        </Drawer.Actions>
      </Drawer>
    </Drawer.Trigger>
  ),
};

export const Controlled: Story = {
  render: args => {
    const [open, setOpen] = useState(false);
    const onOpenChange = (open: boolean) => {
      console.log('open', open);
      setOpen(open);
    };
    return (
      <Stack space={8} alignX="left">
        <Drawer.Trigger open={open} onOpenChange={onOpenChange}>
          <Button>Open Drawer</Button>
          <Drawer {...args}>
            <Drawer.Title>Drawer Title</Drawer.Title>
            <Drawer.Content>Drawer Content</Drawer.Content>
            <Drawer.Actions>
              <Button slot="close">Close</Button>
              <Button variant="primary">Apply</Button>
            </Drawer.Actions>
          </Drawer>
        </Drawer.Trigger>
        <pre>Drawer is {open ? 'open' : 'closed'}</pre>
      </Stack>
    );
  },
};
