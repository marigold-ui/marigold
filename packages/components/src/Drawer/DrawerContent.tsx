import type { ReactNode } from 'react';
import { cn, useClassNames } from '@marigold/system';
import { useDrawerContext } from './Context';

// Props
// ---------------
export interface DrawerContentProps {
  variant?: string;
  size?: string;
  /**
   * Children of the component.
   */
  children?: ReactNode;
}

// Component
// ---------------
export const DrawerContent = ({
  variant,
  size,
  children,
}: DrawerContentProps) => {
  const ctx = useDrawerContext();
  const classNames = useClassNames({
    component: 'Drawer',
    variant: variant ?? ctx.variant,
    size: size ?? ctx.size,
  });

  return (
    <div className={cn('[grid-area:content]', classNames.content)}>
      {children}
    </div>
  );
};
