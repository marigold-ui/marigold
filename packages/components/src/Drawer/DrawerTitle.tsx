import type { CSSProperties } from 'react';
import { Header, Heading } from 'react-aria-components';
import { cn, useClassNames } from '@marigold/system';
import { useDrawerContext } from './Context';

export interface DrawerTitleProps {
  variant?: string;
  size?: string;
  /**
   * Sets heading level, h1 through h6.
   * @default "2"
   */
  level?: '1' | '2' | '3' | '4' | '5' | '6';

  /**
   * Children of the component.
   */
  children?: React.ReactNode;
}

export const DrawerTitle = ({
  level = '2',
  variant,
  size,
  children,
}: DrawerTitleProps) => {
  const ctx = useDrawerContext();
  const classNames = useClassNames({
    component: 'Drawer',
    variant: variant ?? ctx.variant,
    size: size ?? ctx.size,
  });

  return (
    <Header
      className={cn('[grid-area:title]', classNames.header)}
      style={{ '--i': 0 } as CSSProperties}
    >
      <Heading slot="title" level={Number(level)} className={classNames.title}>
        {children}
      </Heading>
    </Header>
  );
};
