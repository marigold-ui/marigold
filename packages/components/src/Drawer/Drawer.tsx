import type { CSSProperties } from 'react';
import { useContext, useRef } from 'react';
import type { DialogProps } from 'react-aria-components';
import { Dialog, OverlayTriggerStateContext } from 'react-aria-components';
import type { AriaLandmarkRole } from '@react-aria/landmark';
import { useLandmark } from '@react-aria/landmark';
import { cn, useClassNames, useSmallScreen } from '@marigold/system';
import { CloseButton } from '../CloseButton';
import { DrawerContext } from './Context';
import { DrawerActions } from './DrawerActions';
import { DrawerContent } from './DrawerContent';
import { DrawerModal } from './DrawerModal';
import { DrawerTitle } from './DrawerTitle';
import { DrawerTrigger } from './DrawerTrigger';

// Props
// ---------------
export interface DrawerProps
  extends Omit<DialogProps, 'className' | 'style' | 'isOpen' | 'role'> {
  size?: 'default' | 'xsmall' | 'small' | 'medium' | (string & {});

  /**
   * Whether the overlay is open by default (controlled).
   * @default undefined
   */
  open?: boolean;

  /**
   * Whether pressing the escape key closes the modal.
   * @default true
   */

  /**
   * The placement of the drawer on the screen.
   * @default right
   */
  placement?: 'top' | 'bottom' | 'left' | 'right';

  keyboardDismissable?: boolean;

  /**
   * Show the close button.
   */
  closeButton?: boolean;

  /**
   * The `role` property sets the ARIA landmark role for this component,
   * enhancing accessibility by clarifying its purpose to assistive technologies.
   *
   * Only ARIA landmark roles (e.g., "complementary", "search", "banner", "navigation")
   * can be used to ensure proper semantic context. Defaults to `"complementary"`
   * for secondary content (e.g., filters, sidebar) that supports the main content.
   * @default "complementary"
   */
  role?: Exclude<AriaLandmarkRole, 'main'>;
}

// Component
// ---------------
export const Drawer = ({
  children,
  size,
  open,
  keyboardDismissable,
  closeButton,
  role = 'complementary',
  placement = 'right',
  ...props
}: DrawerProps) => {
  const ref = useRef<HTMLElement | null>(null);
  const classNames = useClassNames({
    component: 'Drawer',
    variant: placement,
    size,
  });

  const ctx = useContext(OverlayTriggerStateContext);

  /**
   * On smaller screens the we render a modal dialog instead of a non-modal drawer
   * and need to adjust the role and props accordingly.
   */
  const isSmallScreen = useSmallScreen();
  const landmarkAria = useLandmark({ ...props, role }, ref);
  const landmarkProps = isSmallScreen ? {} : landmarkAria.landmarkProps;

  return (
    <DrawerModal
      className={classNames.overlay}
      open={open}
      keyboardDismissable={keyboardDismissable}
    >
      <DrawerContext.Provider value={{ variant: placement, size }}>
        <Dialog
          {...props}
          // Override RAC here so we can set an appropriate role
          {...(landmarkProps as any)}
          className={cn(
            'h-(--visual-viewport-height) outline-none',
            'grid [grid-template-areas:"title"_"content"_"actions"]',
            classNames.container
          )}
        >
          {closeButton && (
            <CloseButton
              aria-label="dismiss drawer"
              style={{ '--i': 0 } as CSSProperties}
              className={classNames.closeButton}
              onPress={ctx?.close}
            />
          )}
          {children}
        </Dialog>
      </DrawerContext.Provider>
    </DrawerModal>
  );
};

Drawer.Trigger = DrawerTrigger;
Drawer.Title = DrawerTitle;
Drawer.Content = DrawerContent;
Drawer.Actions = DrawerActions;
