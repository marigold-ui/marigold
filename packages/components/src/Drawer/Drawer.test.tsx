import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';
import type { Theme } from '@marigold/system';
import { cva } from '@marigold/system';
import { Button } from '../Button';
import { setup } from '../test.utils';
import type { DrawerProps } from './Drawer';
import { Drawer } from './Drawer';

let isSmallScreen = false;
const mockMatchMedia = () =>
  vi.fn().mockImplementation(() => {
    return {
      matches: isSmallScreen,
    };
  });
window.matchMedia = mockMatchMedia();

const theme: Theme = {
  name: 'test',
  components: {
    Button: cva(),
    CloseButton: cva('size-5'),
    Drawer: {
      overlay: cva(),
      container: cva('p-5', {
        variants: {
          variant: {
            custom: 'bg-green-400',
          },
          size: {
            large: 'w-[400px]',
          },
        },
      }),
      closeButton: cva('p-1', {
        variants: {
          variant: {
            custom: 'bg-black',
          },
        },
      }),
      header: cva(),
      title: cva(),
      content: cva(),
      actions: cva(),
    },
    Underlay: cva(),
    Modal: cva(),
  },
};

const Component = (props: DrawerProps) => (
  <Drawer.Trigger>
    <Button>Toggle</Button>
    <Drawer {...props}>
      <Drawer.Title>Drawer Title</Drawer.Title>
      <Drawer.Content>Drawer Content</Drawer.Content>
      <Drawer.Actions>
        <Button slot="close">Close</Button>
        <Button variant="primary">Save</Button>
      </Drawer.Actions>
    </Drawer>
  </Drawer.Trigger>
);

const user = userEvent.setup();
const { render } = setup({ theme });

test('renders nothing if isOpen is not set', () => {
  render(<Component />);
  expect(screen.queryByText('Drawer Content')).toBeNull();
});

test('opens/closes via trigger', async () => {
  render(<Component />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  expect(screen.queryByText('Drawer Content')).not.toBeInTheDocument();

  await user.click(button);

  const drawer = screen.getByText('Drawer Content');
  expect(drawer).toBeInTheDocument();

  await user.click(button);

  expect(drawer).not.toBeInTheDocument();
});

test('can be closed with esc key', async () => {
  render(<Component />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  const drawer = screen.getByText('Drawer Content');
  expect(drawer).toBeInTheDocument();

  await user.keyboard('{Escape}');
  expect(drawer).not.toBeInTheDocument();
});

test('disable closing via esc key', async () => {
  render(<Component keyboardDismissable={false} />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  const drawer = screen.getByText('Drawer Content');
  expect(drawer).toBeInTheDocument();

  await user.keyboard('{Escape}');
  expect(drawer).toBeInTheDocument();
});

test('can be closed via button with [slot="close"]', async () => {
  render(<Component />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  const drawer = screen.getByText('Drawer Content');
  expect(drawer).toBeInTheDocument();

  const close = screen.getByRole('button', { name: 'Close' });
  await user.click(close);

  expect(drawer).not.toBeInTheDocument();
});

test('has "complementary" role by default', async () => {
  render(<Component />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  expect(screen.getByRole('complementary')).toBeInTheDocument();
});

test('allows to set other landmark roles', async () => {
  render(<Component role="navigation" />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  expect(screen.getByRole('navigation')).toBeInTheDocument();
});

test('able to show a close button', async () => {
  render(<Component closeButton />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  expect(screen.getByLabelText('dismiss drawer')).toBeInTheDocument();
});

test('able to close via close button', async () => {
  render(<Component closeButton />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  const drawer = screen.getByText('Drawer Content');
  expect(drawer).toBeInTheDocument();

  const close = screen.getByLabelText('dismiss drawer');
  await user.click(close);

  expect(drawer).not.toBeInTheDocument();
});

test('uses modal on small screens', async () => {
  isSmallScreen = true;
  render(<Component closeButton />);

  const button = screen.getByRole('button', { name: 'Toggle' });
  await user.click(button);

  const drawer = screen.getByText('Drawer Content');
  expect(drawer).toBeInTheDocument();

  expect(screen.getByRole('dialog')).toBeInTheDocument();
  expect(screen.queryByRole('complementary')).not.toBeInTheDocument();
});
