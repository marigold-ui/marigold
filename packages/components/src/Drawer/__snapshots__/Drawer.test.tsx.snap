// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`slides from the left > "entering:animate-slide-in-left exiting:animate-slide-out-left top-0 left-0" 1`] = `"fixed"`;
